File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/Views/ParticipantsView/ParticipantView/ParticipantView.m
//
//  ParticipantView.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 15.08.2021.
//  Copyright © 2021 QuickBlox Team. All rights reserved.
//

#import "ParticipantView.h"
#import "UILabel+Videochat.h"
#import "UIView+Videochat.h"
#import "NSString+Videochat.h"
#import "UIColor+Videochat.h"

@interface ParticipantView()
@end

@implementation ParticipantView
//MARK: - Life Cycle
- (void)awakeFromNib {
    [super awakeFromNib];

    [self setupViews];
}

//MARK - Setup
- (void)setupViews {
    self.backgroundColor = UIColor.clearColor;
    self.containerView.backgroundColor = [UIColor colorWithRed:0.20f green:0.20f blue:0.20f alpha:1.0f];
    [self.userAvatarLabel setRoundedLabelWithCornerRadius:30.0f];
    [self.userAvatarImageView setRoundViewWithCornerRadius:30.0f];
    self.userAvatarImageView.hidden = YES;
    self.nameLabelCenterXConstraint.constant = 0.0f;
    self.callingInfoLabelHeightConstraint.constant = 28.0f;
}

- (void)setName:(NSString *)name {
    _name = name;
    self.nameLabel.text = self.name;
    self.userAvatarLabel.text = name.firstLetter;
}

- (void)setParticipantId:(NSNumber *)userID {
    _participantId = userID;
    self.userAvatarLabel.backgroundColor = [UIColor colorWithHexString:[NSString stringWithFormat:@"#%lX",
                                                                            (unsigned long)self.participantId.unsignedIntValue]];
}

- (void)setIsCallingInfo:(BOOL)isCallingInfo {
    _isCallingInfo = isCallingInfo;
    if (_isCallingInfo == NO) {
        self.callingInfoLabelHeightConstraint.constant = 0.0f;
    } else {
        self.callingInfoLabelHeightConstraint.constant = 28.0f;
    }
}

- (void)setConnectionState:(QBRTCConnectionState)connectionState {
    
    if (_connectionState != connectionState) {
        _connectionState = connectionState;
        
        switch (connectionState) {
            case QBRTCConnectionStateConnected:
                self.stateLabel.text = @"";
                self.callingInfoLabelHeightConstraint.constant = 0.0f;
                self.stateLabel.hidden = YES;
                break;

            case QBRTCConnectionStateClosed:
                self.stateLabel.text = @"Closed";
                [self setupHiddenViews];
                break;

            case QBRTCConnectionStateFailed:
                self.stateLabel.text = @"Failed";
                [self setupHiddenViews];
                break;

            case QBRTCConnectionStateHangUp:
                self.stateLabel.text = @"Hung Up";
                [self setupHiddenViews];
                break;

            case QBRTCConnectionStateRejected:
                self.stateLabel.text = @"Rejected";
                [self setupHiddenViews];
                break;

            case QBRTCConnectionStateNoAnswer:
                self.stateLabel.text = @"No Answer";
                [self setupHiddenViews];
                break;

            case QBRTCConnectionStateDisconnectTimeout:
                self.stateLabel.text = @"Time out";
                [self setupHiddenViews];
                break;

            case QBRTCConnectionStateDisconnected:
                self.stateLabel.text = @"Disconnected";
                [self setupHiddenViews];
                break;

            default:
                self.stateLabel.text = @"";
                break;
        }
    }
}

- (void)setupHiddenViews {
    self.callingInfoLabelHeightConstraint.constant = 0.0f;
    self.stateLabel.hidden = NO;
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/Views/ParticipantView/ParticipantView.swift
//
//  ParticipantView.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 08.06.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import QuickbloxWebRTC

struct ParticipantViewConstant {
    static let stateColor = #colorLiteral(red: 0.700391233, green: 0.7436676621, blue: 0.8309402466, alpha: 1)
}

class ParticipantView: UIView {
    
    //MARK: - IBOutlets
    @IBOutlet weak var userView: UIView!
    @IBOutlet weak var userAvatarLabel: UILabel!
    @IBOutlet weak var userAvatarImageView: UIImageView!
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var containerView: UIView!
    @IBOutlet weak var nameLabelCenterXConstraint: NSLayoutConstraint!
    @IBOutlet weak var stateLabel: UILabel!
    @IBOutlet weak var callingInfoLabelHeightConstraint: NSLayoutConstraint!

    
    //MARK: - Properties
    var name = "" {
        didSet {
            nameLabel.text = name
            userAvatarLabel.text = String(name.capitalized.first ?? Character("Q"))
        }
    }
    
    var isCallingInfo = true {
        didSet {
            if isCallingInfo == false {
                callingInfoLabelHeightConstraint.constant = 0.0
            } else {
                callingInfoLabelHeightConstraint.constant = 28.0
            }
        }
    }
    
    var ID: UInt = 0 {
        didSet {
            userAvatarLabel.backgroundColor = ID.generateColor()
        }
    }
    
    var connectionState: QBRTCConnectionState = .connecting {
        didSet {
            switch connectionState {
            case .connected:
                stateLabel.text = ""
                callingInfoLabelHeightConstraint.constant = 0.0
                stateLabel.isHidden = true
            case .closed:
                stateLabel.text = "Closed"
                setupHiddenViews()
            case .failed:
                stateLabel.text = "Failed"
                setupHiddenViews()
            case .hangUp:
                stateLabel.text = "Hung Up"
                setupHiddenViews()
            case .rejected:
                stateLabel.text = "Rejected"
                setupHiddenViews()
            case .noAnswer:
                stateLabel.text = "No Answer"
                setupHiddenViews()
            case .disconnectTimeout:
                stateLabel.text = "Time out"
                setupHiddenViews()
            case .disconnected:
                stateLabel.text = "Disconnected"
                setupHiddenViews()
            default: stateLabel.text = ""
            }
        }
    }
    
    //MARK - Setup
    func setupViews() {
        backgroundColor = .clear
        containerView.backgroundColor = #colorLiteral(red: 0.1960526407, green: 0.1960932612, blue: 0.1960500479, alpha: 1)
        userAvatarLabel.setRoundedLabel(cornerRadius: 30.0)
        userAvatarImageView.setRoundedView(cornerRadius: 30.0)
        userAvatarImageView.isHidden = true
        nameLabelCenterXConstraint.constant = 0.0
        callingInfoLabelHeightConstraint.constant = 28.0
    }
    
    func setupHiddenViews() {
        callingInfoLabelHeightConstraint.constant = 0.0
        stateLabel.isHidden = false
    }
    
    //MARK: - Life Cycle
    override func awakeFromNib() {
        super.awakeFromNib()
        
        setupViews()
    }
}
